proposals/0339-increase-cpi-account-info-limit.md

```diff
@@ -0,0 +1,108 @@
+---
+simd: '0339'
+title: Increase CPI Account Info Limit
+authors:
+  - Justin Starry (Anza)
+category: Standard
+type: Core
+status: Review
+created: 2025-08-15
+feature: (fill in with feature key and github tracking issues once accepted)
+---
+
+## Summary
+
+Increase the maximum account info length for cross-program invoked (CPI)
+instructions from 64 to 255 and consume compute units for serialized account
+info's and instruction account meta's.
+
+## Motivation
+
+CPI's are restricted to a limit of 64 account info's passed to the syscall.
+This limit is burdensome for onchain programs which themselves were invoked with
+more than 64 accounts because it means they cannot simply pass through the same
+list of account info's that they were invoked with to another CPI syscall. They
+are faced with the burden of first deduplicating the account info's and
+constructing a new list before making the syscall.
+
+This problem arises when onchain programs wrap another program (such as Jupiter)
+that composes many other programs and are invoked with over 64 accounts
+(including duplicates).
+
+Along with the increasing the account info limit, start charging for both
+serialized account info's and instruction account meta's to avoid abuse of
+resources in CPI calls.
+
+## New Terminology
+
+- **Account Info:** The serialized information for each account referenced in a
+CPI instruction used to read account info from the caller program.
+
+## Detailed Design
+
+### Maximum Account Info Length
+
+Increase the maximum account info length imposed on CPI syscalls from **64** to
+**255**. The maximum is inclusive, meaning that a list of account info's with a
+length of 255 is valid.
+
+### Account Info Cost
+
+Consume **1 compute unit (CU)** for every `cpi_bytes_per_unit` (currently 250)
+bytes of account info.
+
+Fixed size of **80 bytes** for each `account_info` broken down as
+
+  - 32 bytes for account address
+  - 32 bytes for owner address
+  - 8 bytes for lamport balance
+  - 8 bytes for data length
+
+The total cost of account info's can be computed with
+`(account_info_size * account_infos_len) / cpi_bytes_per_unit` rounded down to
```

Comment:
Just to confirm whether the cost is for each unique `account info` or not. This is not very clear from the equation here. For example, if I use a slice of 254 account infos, but all of them are copied from the same one, then the runtime has only to serialize a single account payload.